src/evm/protocol/uniswap_v4/state.rs

```diff
@@ -282,22 +286,117 @@ impl ProtocolSim for UniswapV4State {
             SimulationError::InvalidInput("I256 overflow: amount_in".to_string(), None)
         })?;
 
-        let result = self.swap(zero_for_one, amount_specified, None)?;
+        let mut amount_to_swap = amount_specified;
+        let mut lp_fee_override: Option<u32> = None;
+        let mut before_swap_gas = 0u64;
+        let mut after_swap_gas = 0u64;
+
+        let token_in_address = Address::from_slice(&token_in.address);
+        let token_out_address = Address::from_slice(&token_out.address);
+
+        let state_context = StateContext {
+            currency_0: if zero_for_one { token_in_address } else { token_out_address },
+            currency_1: if zero_for_one { token_out_address } else { token_in_address },
+            fees: self.fees.clone(),
+            tick: self.tick,
+        };
+
+        let swap_params = SwapParams {
+            zero_for_one,
+            amount_specified,
+            sqrt_price_limit: U256::ZERO, // Will be set to appropriate limit in swap
+        };
+
+        // Check if hook is set and has before_swap permissions
+        if let Some(ref hook) = self.hook {
+            if has_permission(hook.address(), HookOptions::BeforeSwap) {
+                let before_swap_params = BeforeSwapParameters {
+                    context: state_context.clone(),
+                    // TODO is this the right sender? Does this matter?
+                    sender: Address::ZERO,
+                    swap_params: swap_params.clone(),
+                    // TODO what is the hook data?
+                    hook_data: Bytes::new(),
+                };
+
+                // TODO use block 0 for now - not sure where to get the actual block number
+                let before_swap_result = hook
+                    .before_swap(before_swap_params, 0)
+                    .map_err(|e| {
+                        SimulationError::FatalError(format!("BeforeSwap hook failed: {e:?}"))
+                    })?;
+
+                before_swap_gas = before_swap_result.gas_estimate;
+                let before_swap_amount_delta = before_swap_result.result.amountDelta;
+
+                // Convert amountDelta to amountToSwap as per Uniswap V4 spec
+                // See: https://github.com/Uniswap/v4-core/blob/main/src/libraries/Hooks.sol#L270
+                if before_swap_amount_delta != I256::ZERO {
+                    amount_to_swap = amount_specified + before_swap_amount_delta;
+                    // TODO in the USV4 code they check if it's exact input or output, and makes
+                    // sure this doesn't change (returns an error if it does). Is that necessary in
+                    // our case, or do we just support exact input so this doesn't matter?
+                }
```

Comment:
huumm is it a big thing? I would say let's add it anyway. this swap method doesn't care which one it is